devdoc/thandle_tuple_array_requirements.md

```diff
@@ -0,0 +1,84 @@
+`thandle_tuple_array` requirements
+================
+
+## Overview
+
+`thandle_tuple_array` holds an array of structs with `THANDLE(<TYPE>)` members along with the count of array elements. That is, an array of tuples where each tuple field is a `THANDLE`.
+
+## Exposed API
+
+```c
+#define DECLARE_THANDLE_TUPLE_TYPE(name, ...) \
+    ...
+
+#define DECLARE_THANDLE_TUPLE_ARRAY_TYPE(name) \
+    ...
+
+#define THANDLE_TUPLE_ARRAY_CREATE(name) ...
+#define THANDLE_TUPLE_ARRAY_DESTROY(name) ...
+
+#define DECLARE_THANDLE_TUPLE_ARRAY(name, ...) \
+    DECLARE_THANDLE_TUPLE_TYPE(name, __VAR_ARGS__) \
+    DECLARE_THANDLE_TUPLE_ARRAY_TYPE(name) \
+    MOCKABLE_FUNCTION(, THANDLE_TUPLE_ARRAY_TYPE(name)*, THANDLE_TUPLE_ARRAY_CREATE(name), uint32_t, count); \
+    MOCKABLE_FUNCTION(, void, THANDLE_TUPLE_ARRAY_DESTROY(name), THANDLE_TUPLE_ARRAY_TYPE(name)*, tuple_array);
+
+#define DEFINE_THANDLE_TUPLE_ARRAY(name, ...) \
+    ...
+```
+
+### Example Usage
+
+```c
+// In the header
+#define MY_TUPLE_FIELDS \
+    RC_STRING, foo, \
+    RC_STRING, bar, \
+    RC_STRING, baz
+
+DECLARE_THANDLE_TUPLE_ARRAY(MY_TUPLE, MY_TUPLE_FIELDS);

```

Comment:
>DECLARE_THANDLE_TUPLE_ARRAY [](http://example.com/codeflow?start=0&length=27)

It took me a while to understand what these are. 

I do need to ask why isn't this broken in 2 separate concerns:

Concept 1: have a THANDLE'd struct - this is an independent data type that is like-a-C-struct, except that ALL fields are THANDLEs and the structure itself is a THANDLE.

Example: (codeflow cam compile anything)

DECLARE_THANDLE_STRUCT(FIELDS_S, MY_TUPLE_FIELDS).

DEFINE_THANDLE_STRUCT(FIELDS_S, MY_TUPLE_FIELDS).

... and then have the usual THANDLE macros working on the type (allegedly - somehow).

Concept2: have a THANDLE'd array - this another data type that is-like-C-array and its base type is THANDLE and the complete array is THANDLE. So this new THANDLE array could have its individual elements as either "basic" THANDLEs or structs or other THANDLE_ARRAY. 

... I am curious.

Reply:
hmm, well to give some context, today we have `rc_string_array`. It's just like this but it's a wrapper of `THANDLE(RC_STRING)`'s in an array (so array of thandle + count).

I need to instead pass out an array of 2 strings. So I wanted to generalize something better than `rc_string_array_of_pairs`. I could have made it just be an arbitrary number of strings, but it's easy to use any THANDLE and not limit it to strings.

Now I guess some of this could the generalized further to THANDLE, but it wasn't something I needed in my use case. I'm tempted to just keep it as is to meet my needs because I don't need any additional ref counting really (I just want to have a nice way to pass the array and have a destroy function so the caller doesn't need to do that)

Reply:
I added a note about this, saying that this could be changed to be a THANDLE as an improvement.